File: Python_version/examples/ex4_osqp/combinedFunction.py
import numpy as np
from numpy import linalg as LA
from numpy.core.fromnumeric import transpose
import os, sys
currentdir = os.path.dirname(os.path.realpath(__file__))
import scipy.io
import scipy.linalg as la
from numpy import conjugate as conj

class f_gradStruct:
    pass

class ciStruct:
    pass

class ci_gradStruct:
    class c1:
        pass
    class c2:
        pass

def combinedFunction(X,parameters):
    
    # input variable, matirx form
    XX           = X.XX

    # read input data from matlab file
    # file = currentdir + "/ex4_data_n=200.mat"
    # mat = scipy.io.loadmat(file)
    # mat_struct = mat['sys']
    # val = mat_struct[0,0]
    # A = val['A']
    # B = val['B']
    # p = B.shape[1]
    # C = val['C']
    # m = C.shape[0]

    A = parameters.A
    B = parameters.B
    C = parameters.C
    m = parameters.m
    p = parameters.p

    stability_margin = 1
    M           = A + B@XX@C
    [D,V]       = la.eig(M)
    d           = D
    [D_conj,VL] = la.eig(conj(M.T))
    dl          = conj(D_conj)

    #  OBJECTIVE VALUE AT X
    #  Get the max imaginary part, and an eigenvalue associated with it,
    #  since the constraint is to limit eigenvalues to a band centered on
    #  the x-axis
    d_lst = list(np.imag(d))
    mi   = np.max(d_lst)
    indx        = d_lst.index(mi)
    lambda_var      = d[indx]
    f           = mi

############################################################################
    #  GRADIENT OF THE OBJECTIVE AT X
    #  Get its corresponding right eigenvector
    x           = (V[:,indx]).reshape(p*m,1)
    #  Now find the matching left eigenvector for lambda
    dl_lst = list(np.abs(dl - lambda_var))
    indx    = dl_lst.index(np.min(dl_lst))
    y           = (VL[:,indx]).reshape(p*m,1)
    Bty         = (B.T @ y).reshape(p,1)
    Cx          = (C @ x).reshape(m,1)
    #  Gradient of inner product with respect to A
    #  f_grad.XX is a real-valued matrix, p by m
    f_grad = f_gradStruct()
    f_grad.XX      = np.imag((conj(Bty)@Cx.T)/(conj(y.T)@x))
    # print(sum(sum(f_grad.XX)))
    

    
    #  INEQUALITY CONSTRAINT AT X
    #  Compute the spectral abscissa of A from the spectrum and an
    #  eigenvalue associated with it
    ci = ciStruct()
    d_lst = list(np.real(d))
    ci.c1   = max(d_lst)
    indx        = d_lst.index(ci.c1)
    lambda_var      = d[indx]
    #  account for the stability margin in the inequality constraint
    ci.c1          = ci.c1 + stability_margin



    #  GRADIENT OF THE INEQUALITY CONSTRAINT AT X
    #  Get its corresponding right eigenvector
    ci_grad = ci_gradStruct()
    x           = (V[:,indx]).reshape(p*m,1)
    #  Now find the matching left eigenvector for lambda
    dl_lst = list(abs(dl - lambda_var))
    indx    = dl_lst.index(min(dl_lst))
    y           = (VL[:,indx]).reshape(p*m,1)
    Bty         = (B.T @ y).reshape(p,1)
    Cx          = (C @ x).reshape(m,1)
    #  Gradient of inner product with respect to A
    ci_grad.c1.XX     = np.real((conj(Bty)@Cx.T)/(conj(y.T) @ x))
    # print(sum(sum(ci_grad.c1.XX)))
    ci_grad.c1.XX = (ci_grad.c1.XX).reshape(p*m,1)

    
    #  EQUALITY CONSTRAINT
    #  Return [] when (in)equality constraints are not present.
    ce = None
    ce_grad = None

    return [f,f_grad,ci,ci_grad,ce,ce_grad]

File: tobedelete/MATLAB_version/example/ex4/combinedFunction.m
function [f,f_grad,ci,ci_grad,ce,ce_grad] = combinedFunction(X, parameters)
%   combinedFunction: (example_mat/ex2_spectral_radius_optimization)
%       Defines objective and inequality constraint functions, with their
%       respective gradients, for an eigenvalue optimization problem of a 
%       static-output-feedback (SOF) plant:
%       
%           M = A + BXC,
%
%       where A,B,C are all fixed real-valued matrices
%           - A is n by n 
%           - B is n by p
%           - C is m by n
%       and X is a real-valued p by m matrix of the optimization variables.
%
%   USAGE:
%       [f,f_grad,ci,ci_grad,ce,ce_grad] = ...
%                               combinedFunction(X);
%
%   INPUT:
%       X                   struct of optimization variables
%                           X.XX is a real-valued matrix, p by m
%
%   OUTPUT:
%       f                   value of the objective function at x
%                           scalar real value
%
%       f_grad              struct of gradient of the objective function at x.
%                           f_grad.XX is a real-valued matrix, p by m 
%
%       ci                  struct of value of the inequality constraint at x.
%                           c.c1 is the value of the inequality constraint at x.
%                           scalar real value
%
%       ci_grad             struct of gradient of the inequality constraint at x.
%                           c.c1.XX is a real-valued matrix,  p*m by 1 
%
%       ce                  value of the equality constraint(s) at x.
%                           [], since there aren't any equality constraints
%
%       ce_grad             gradient(s) of the equality constraint(s) at x.
%                           [], since there aren't any equality constraints
%
%

XX           = X.XX;

% [A,B,C,~,~] = loadExample('ex4_data_n=200.mat');
A = parameters.A;
B = parameters.B;
C = parameters.C;
stability_margin = 1;
M           = A + B*XX*C;
[V,D]       = eig(M);
d           = diag(D);
[VL,D_conj] = eig(M');
dl          = conj(diag(D_conj));

% OBJECTIVE VALUE AT X
% Get the max imaginary part, and an eigenvalue associated with it,
% since the constraint is to limit eigenvalues to a band centered on
% the x-axis
[mi,indx]   = max(imag(d));
indx        = indx(1);
lambda      = d(indx);
f           = mi;

% GRADIENT OF THE OBJECTIVE AT X
% Get its corresponding right eigenvector
x           = V(:,indx);
% Now find the matching left eigenvector for lambda
[~,indx]    = min(abs(dl - lambda));
y           = VL(:,indx);
Bty         = B.'*y;
Cx          = C*x;
% Gradient of inner product with respect to A
% f_grad.XX is a real-valued matrix, p by m
f_grad.XX      = imag((conj(Bty)*Cx.')/(y'*x));



% INEQUALITY CONSTRAINT AT X
% Compute the spectral abscissa of A from the spectrum and an
% eigenvalue associated with it
[ci.c1,indx]   = max(real(d));
indx        = indx(1);
lambda      = d(indx);
% account for the stability margin in the inequality constraint
ci.c1          = ci.c1 + stability_margin;

% GRADIENT OF THE INEQUALITY CONSTRAINT AT X
% Get its corresponding right eigenvector
x           = V(:,indx);
% Now find the matching left eigenvector for lambda
[~,indx]    = min(abs(dl - lambda));
indx        = indx(1);
y           = VL(:,indx);
Bty         = B.'*y;
Cx          = C*x;
% Gradient of inner product with respect to A
ci_grad.c1.XX     = real((conj(Bty)*Cx.')/(y'*x));
ci_grad.c1.XX = ci_grad.c1.XX(:);

% EQUALITY CONSTRAINT
% Return [] when (in)equality constraints are not present.
ce = [];
ce_grad=[];
end

function [A,B,C,p,m] = loadExample(filename)
% MATLAB does not allow loading variable names into a function scope
% that has nested functions so we need a completely separate function
% for this task.
load(filename);
A   = sys.A;
B   = sys.B;
C   = sys.C;
p   = size(B,2);
m   = size(C,1);
end